homeassistant/components/legrandinone.py

```diff
@@ -0,0 +1,505 @@
+        connection = create_iobl_connection(
+            port=port,
+            host=host,
+            event_callback=event_callback,
+            disconnect_callback=reconnect,
+            loop=hass.loop,
+            ignore=None
+        )
+
+        try:
+            with async_timeout.timeout(CONNECTION_TIMEOUT,
+                                       loop=hass.loop):
+                transport, protocol = await connection
+
+        except (serial.serialutil.SerialException, ConnectionRefusedError,
+                TimeoutError, OSError, asyncio.TimeoutError) as exc:
+            reconnect_interval = config[DOMAIN][CONF_RECONNECT_INTERVAL]
+            _LOGGER.exception(
+                "Error connecting to iobl, reconnecting in %s",
+                reconnect_interval)
+            # Connection to IOBL device is lost, make entities unavailable
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+            hass.loop.call_later(reconnect_interval, reconnect, exc)
+            return
+
+        # There is a valid connection to a IOBL device now so
+        # mark entities as available
+        async_dispatcher_send(hass, SIGNAL_AVAILABILITY, True)
+
+        # Bind protocol to command class to allow entities to send commands
+        LegrandInOneCommand.set_iobl_protocol(protocol)
+
+        # handle shutdown of IOBL asyncio transport
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                   lambda x: transport.close())
+
+        _LOGGER.info('Connected to Iobl')
+
+    hass.async_create_task(connect())
+    return True
+
+
+class LegrandInOneDevice(Entity):
+    """Representation of an iobl device.
+
+    Contains the common logic for Iobl entities.
+    """
+
+    platform = None
+    _state = None
+    _available = True
+    legrand_id = None
+    iobl_mode = None
+    iobl_media = None
+    iobl_type = None
+    iobl_unit = None
+
+    def __init__(self, device_id, initial_event=None, name=None,
+                 fire_event=False, iobl_media='plc', iobl_comm_mode='unicast'):
+        """Initialize the device."""
+        self._initial_event = initial_event
+        self.legrand_id = device_id
+        self.iobl_mode = iobl_comm_mode
+        self.iobl_media = iobl_media
+
+        if name:
+            self._name = name
+        else:
+            self._name = device_id
+
+        self._should_fire_event = fire_event
+
+    @callback
+    def handle_event_callback(self, event):
+        """Handle incoming event for device type."""
+        # Call platform specific event handler
+        self._handle_event(event)
+
+        # Propagate changes through ha
+        self.async_schedule_update_ha_state()
+
+        # Put command onto bus for user to subscribe to
+        if self._should_fire_event and identify_event_type(
+                event) == EVENT_KEY_COMMAND:
+            self.hass.bus.async_fire(EVENT_BUTTON_PRESSED, {
```

Comment:
Entities shouldn't fire custom events. Entities already spawn state changed events. If a device needs to fire events it should do some from the legrandinone component.

Reply:
Ok. I remove that. It was an RFLink legacy.